src/__tests__/typeChecker.test.ts

```diff
@@ -0,0 +1,311 @@
+      }
+      // const xs4 = remove(true, xs3);
+    `
+    // TODO redo how we test for this
+    // const program = typeCheck(parse(code1, 2))
+    // // @ts-ignore
+    // expect(program.body[2].declarations[0].init.inferredType).toEqual<Type>({
+    //   kind: 'primitive',
+    //   name: 'number'
+    // })
+    const [program, errors] = typeCheck(parse(code1, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+"accumulate: ((number, number) -> number, number, [number, List<number>]) -> number
+xs: [number, List<number>]
+y: number
+map: Couldn't infer type
+xs1: [boolean, List<boolean>]
+xs2: [boolean, List<boolean>]
+append: Couldn't infer type
+xs3: [boolean, List<boolean>]
+remove: Couldn't infer type"
+`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('works for accumulate used with different kinds of pairs', () => {
+    const code = `
+      function accumulate(op, init, xs) {
+        return is_null(xs) ? init : op(head(xs), accumulate(op, init, tail(xs)));
+      }
+      const xs = pair(1, pair(2, null));
+      const ys = pair(true, pair(true, null));
+      accumulate((x,y)=>x+y,0,xs);
+      accumulate((x,y)=>x||y,0,ys);
+    `
+    const [program, errors] = typeCheck(parse(code, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+"accumulate: ((T67, T21) -> T21, T21, [T67, List<T67>]) -> T21
+xs: [number, List<number>]
+ys: [boolean, List<boolean>]"
+`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('Will not work if used in a monomorphic manner', () => {
+    const code = `
+      function accumulate(op, init, xs) {
+        return is_null(xs) ? init : op(head(xs), accumulate(op, init, tail(xs)));
+      }
+      const xs = pair(1, pair(2, null));
+      const ys = pair(true, pair(true, null));
+      const a = accumulate((x,y)=>x+y,0,xs);
+      const b = accumulate((x,y)=>x||y,0,ys);
+    `
+
+    const [program, errors] = typeCheck(parse(code, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+"accumulate: ((number, number) -> number, number, [number, List<number>]) -> number
+xs: [number, List<number>]
+ys: [boolean, List<boolean>]
+a: number
+b: number"
+`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+"Line 8: Types do not unify: number vs boolean
+Line 8: Types do not unify: boolean vs number"
+`)
+  })
+})
+
+describe('type checking functions', () => {
+  it('happy paths for recursive functions', () => {
+    const code1 = `
+      function append(xs, ys) {
+        return is_null(xs) ? ys : pair(head(xs), append(tail(xs), ys));
+      }
+    `
+    const [program, errors] = typeCheck(parse(code1, 2))
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(
+      `"append: ([T29, List<T29>], [T29, List<T29>]) -> [T29, List<T29>]"`
+    )
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('unhappy paths for recursive functions', () => {
```

Comment:
What does it mean to be unhappy?

Reply:
It simply refers to recursive functions being defined/used in a way that it will throw a type error when using our checker.

Reply:
unhappy sounds intuitive enough.